text/0000-variadic.md

```diff
@@ -0,0 +1,244 @@
+only partial solutions for a few target architectures.
+
+This RFC does not propose an interface intended for native Rust code to pass
+variable numbers of arguments to a native Rust function, nor an interface that
+provides any kind of type safety. This proposal exists primarily to allow Rust
+to provide interfaces callable from C code.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+C code allows declaring a function callable with a variable number of
+arguments, using an ellipsis (`...`) at the end of the argument list. For
+compatibility, unsafe Rust code may export a function compatible with this
+mechanism.
+
+Such a declaration looks like this:
+
+```rust
+pub unsafe extern "C" fn func(arg: T, arg2: T2, mut args: ...) {
+    // implementation
+}
+```
+
+The use of `...` as the type of `args` at the end of the argument list declares
+the function as variadic. This must appear as the last argument of the
+function.  The function must use `extern "C"`, and must use `unsafe`. To expose
+such a function as a symbol for C code to call directly, the function may want
+to use `#[no_mangle]` as well; however, Rust code may also pass the function to
+C code expecting a function pointer to a variadic function.
+
+The `args` named in the function declaration has the type
+`core::intrinsics::VaList<'a>`, where the compiler supplies a lifetime `'a`
+that prevents the arguments from outliving the variadic function.
+
+To access the arguments, Rust provides the following public interfaces in
+`core::intrinsics` (also available via `std::intrinsics`):
+
+```rust
+/// The argument list of a C-compatible variadic function, corresponding to the
+/// underlying C `va_list`. Opaque.
+pub struct VaList<'a>;
+
+impl<'a> VaList<'a> {
+    /// Extract the next argument from the argument list.
+    pub unsafe fn arg<T: VaArg>(&mut self) -> T;
+}
+
+impl<'a> Clone for VaList<'a>;
+impl<'a> Drop for VaList<'a>;
+
+/// The type of arguments extractable from VaList
+unsafe trait VaArg;
+
+unsafe impl VaArg for i8;
+unsafe impl VaArg for i16;
+unsafe impl VaArg for i32;
+unsafe impl VaArg for i64;
+unsafe impl VaArg for isize;
+
+unsafe impl VaArg for u8;
+unsafe impl VaArg for u16;
+unsafe impl VaArg for u32;
+unsafe impl VaArg for u64;
+unsafe impl VaArg for usize;
+
+unsafe impl VaArg for f32;
+unsafe impl VaArg for f64;
+
+unsafe impl<T> VaArg for *const T;
+unsafe impl<T> VaArg for *mut T;
+```
+
+All of the corresponding C integer and float types defined in the `libc` crate
+consist of aliases for the underlying Rust types, making it unnecessary for
+`libc` to provide additional implementations of the `VaArg` trait. Nothing
+outside of `core` should define any implementation of `VaArg`.
+
+Note that extracting an argument from a `VaList` follows the platform-specific
+rules for argument passing and promotion. In particular, many platforms promote
+any argument smaller than a C `int` to an `int`. On such platforms, extracting
```

Comment:
This is true for all platforms, it's required by the C standard.

> [If the argument has integral or enumeration type that is subject to the integral promotions, or a floating-point type that is subject to the floating-point promotion, the value of the argument is converted to the promoted type before the call.][promotion]

Basically, any integral type smaller than an `int` is converted to `int` (and any floating point type smaller than a `double` (basically `float` or `short float`, for implementations that provide the latter, is converted to `double`)).

Side note, this paragraph points out something important:

> [After these conversions, if the argument does not have arithmetic, enumeration, pointer, pointer to member, or class type, the program is ill-formed.][allowed-types]

emphasis on `class type` and `enumeration` - this means that it should be valid for programs to impl `VaArg` for their `#[repr(C|uN|iN)]` structure and simple enumeration types.

I'd also argue that `VaArg` should be implemented for `&T`, `&mut T`, `Option<&T>`, and `Option<&mut T>`. This is all unsafe anyways :)

[promotion]: http://eel.is/c++draft/expr.call#9.sentence-6
[allowed-types]: http://eel.is/c++draft/expr.call#9.sentence-4

Reply:
I'll phrase the bit about promotions more carefully, and put some thought into the structure and enum cases.

I'd prefer not to implement `VaArg` for references; there are plenty of ways to convert a pointer to a reference in unsafe code, if you really want a reference, and those ways allow you to inject the appropriate lifetime of that reference more easily.

Reply:
I prefer not treating references specially - they are just pointers after all. And it just makes sense sometimes - if you're implementing `printf`:

```
#[no_mangle]
unsafe extern "C" printf(fmt: &CStr, args: ...) -> c_int {
  for ch in fmt {
    // ...
    // find a %s:
    {
      print!("{}", args.next::<&CStr>());
    }
  }
}
```

Reply:
Except `&CStr` is a fat pointer.

Reply:
@jethrogb for now.

Reply:
@ubsan If, at some point in the future, there's a reasonable possibility of extracting a reference directly from a variadic function argument and getting a reasonable result, then we can add such a mechanism at that time. Until then, though, I'd like to stick with only allowing raw pointers.

Reply:
@ubsan I clarified the explanation of promotion to take your explanation into account; thanks.

Reply:
@joshtriplett I disagree with the reasoning there. There is no reason not to allow references. You almost never actually want raw pointers when dealing with C; you want `Option<&T>`.